File: 4. Estructuras de datos lineales/JavaScript/lista_enlazada_doble.js
// 1. Implementa una lista enlazada doble con funciones para agregar y eliminar nodos en posiciones específicas.
// 2. Escribe una función que devuelva el valor del nodo central en una lista enlazada doble.

class Nodo {
  constructor(valor) {
    this.valor = valor;
    this.siguiente = null;
    this.anterior = null;
  }
}

class ListaEnlazadaDoble {
  constructor() {
    this.cabeza = null;
    this.cola = null;
    this.longitud = 0;
  }

  agregarNodo(valor) {
    const nuevoNodo = new Nodo(valor);
    if (!this.cabeza) {
      this.cabeza = nuevoNodo;
      this.cola = nuevoNodo;
    } else {
      this.cola.siguiente = nuevoNodo;
      nuevoNodo.anterior = this.cola;
      this.cola = nuevoNodo;
    }
    this.longitud++;
  }

  eliminarNodo(posicion) {
    if (posicion < 0 || posicion >= this.longitud) {
      return null;
    }
    let actual = this.cabeza;
    if (posicion === 0) {
      this.cabeza = actual.siguiente;
      if (this.cabeza) {
        this.cabeza.anterior = null;
      } else {
        this.cola = null;
      }
    } else if (posicion === this.longitud - 1) {
      actual = this.cola;
      this.cola = actual.anterior;
      this.cola.siguiente = null;
    } else {
      let indice = 0;
      let anterior = null;
      while (indice < posicion) {
        anterior = actual;
        actual = actual.siguiente;
        indice++;
      }
      anterior.siguiente = actual.siguiente;
      actual.siguiente.anterior = anterior;
    }
    this.longitud--;
    return actual.valor;
  }

  obtenerNodoCentral() {
    if (!this.cabeza) {
      return null;
    }
    let lento = this.cabeza;
    let rapido = this.cabeza;
    while (rapido && rapido.siguiente) {
      lento = lento.siguiente;
      rapido = rapido.siguiente.siguiente;
    }
    return lento.valor;
  }
}

const listaEnlazadaDoble = new ListaEnlazadaDoble();
listaEnlazadaDoble.agregarNodo(1);
listaEnlazadaDoble.agregarNodo(2);
listaEnlazadaDoble.agregarNodo(3);
listaEnlazadaDoble.agregarNodo(4);
listaEnlazadaDoble.agregarNodo(5);
console.log(listaEnlazadaDoble.obtenerNodoCentral()); // 3

File: 4. Estructuras de datos lineales/Python/lista_enlazada_doble.py
# 1. Implementa una lista enlazada doble con métodos para agregar nodos al inicio y al final.
# 2. Escribe una función que recorra la lista en ambos sentidos (adelante y atrás).


class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.siguiente = None
        self.anterior = None


class ListaEnlazadaDoble:
    def __init__(self):
        self.cabeza = None
        self.cola = None

    def agregar_inicio(self, valor):
        nuevo = Nodo(valor)
        if self.cabeza is None:
            self.cabeza = nuevo
            self.cola = nuevo
        else:
            nuevo.siguiente = self.cabeza
            self.cabeza.anterior = nuevo
            self.cabeza = nuevo

    def agregar_final(self, valor):
        nuevo = Nodo(valor)
        if self.cola is None:
            self.cabeza = nuevo
            self.cola = nuevo
        else:
            nuevo.anterior = self.cola
            self.cola.siguiente = nuevo
            self.cola = nuevo

    def recorrer_adelante(self):
        actual = self.cabeza
        while actual:
            print(actual.valor)
            actual = actual.siguiente

    def recorrer_atras(self):
        actual = self.cola
        while actual:
            print(actual.valor)
            actual = actual.anterior


lista = ListaEnlazadaDoble()
lista.agregar_inicio(1)
lista.agregar_inicio(2)
lista.agregar_inicio(3)
lista.agregar_final(4)
lista.agregar_final(5)
lista.agregar_final(6)

print("Recorrido adelante:")
lista.recorrer_adelante()

print("Recorrido atrás:")
lista.recorrer_atras()
